undefined error

scheduler/storage/mongo.go:77:12: undefined: bson.NewDocument
scheduler/storage/mongo.go:77:29: undefined: bson.EC
scheduler/storage/mongo.go:111:12: undefined: bson.NewDocument
scheduler/storage/mongo.go:111:29: undefined: bson.EC
scheduler/storage/mongo.go:138:11: undefined: bson.NewDocument

Your fix doesn't work :(

`[fbettag@w0rk ~/go/src/my/go-mod-bla] $ go get github.com/rakanalh/scheduler
go: finding github.com/rakanalh/scheduler latest
go: downloading github.com/rakanalh/scheduler v0.0.0-20181003095030-d081b7053e0d
go: finding github.com/mongodb/mongo-go-driver/mongo latest
go: finding github.com/mongodb/mongo-go-driver/bson latest
go: finding github.com/mongodb/mongo-go-driver v0.1.0
go: downloading github.com/mongodb/mongo-go-driver v0.1.0
go: finding github.com/golang/snappy latest
go: finding github.com/xdg/stringprep v1.0.0
go: finding github.com/xdg/scram latest
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/xdg/stringprep v1.0.0
# github.com/rakanalh/scheduler/storage
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:77:12: undefined: bson.NewDocument
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:77:29: undefined: bson.EC
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:111:12: undefined: bson.NewDocument
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:111:29: undefined: bson.EC
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:138:11: undefined: bson.NewDocument
zsh: exit 2     go get github.com/rakanalh/scheduler

[fbettag@w0rk ~/go/src/my/go-mod-bla] $ go get github.com/rakanalh/scheduler/...
go: finding github.com/rakanalh/scheduler latest
# github.com/rakanalh/scheduler/storage
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:77:12: undefined: bson.NewDocument
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:77:29: undefined: bson.EC
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:111:12: undefined: bson.NewDocument
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:111:29: undefined: bson.EC
../../../../pkg/mod/github.com/rakanalh/scheduler@v0.0.0-20181003095030-d081b7053e0d/storage/mongo.go:138:11: undefined: bson.NewDocument
zsh: exit 2     go get github.com/rakanalh/scheduler/...
`

--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,4 @@
 *.db
 *.out
+vendor
+coverage.txt

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -2,44 +2,186 @@
 
 
 [[projects]]
+  digest = "1:0deddd908b6b4b768cfc272c16ee61e7088a60f7fe2f06c547bd3d8e1f8b8e77"
   name = "github.com/davecgh/go-spew"
   packages = ["spew"]
-  revision = "346938d642f2ec3594ed81d874461961cd0faa76"
-  version = "v1.1.0"
+  pruneopts = ""
+  revision = "8991bc29aa16c548c550c7ff78260e27b9ab7c73"
+  version = "v1.1.1"
 
 [[projects]]
+  digest = "1:a01080d20c45c031c13f3828c56e58f4f51d926a482ad10cc0316225097eb7ea"
+  name = "github.com/go-stack/stack"
+  packages = ["."]
+  pruneopts = ""
+  revision = "2fee6af1a9795aafbe0253a0cfbdf668e1fb8a9a"
+  version = "v1.8.0"
+
+[[projects]]
+  branch = "master"
+  digest = "1:2a5888946cdbc8aa360fd43301f9fc7869d663f60d5eedae7d4e6e5e4f06f2bf"
+  name = "github.com/golang/snappy"
+  packages = ["."]
+  pruneopts = ""
+  revision = "2e65f85255dbc3072edf28d6b5b8efc472979f5a"
+
+[[projects]]
+  digest = "1:29145d7af4adafd72a79df5e41456ac9e232d5a28c1cd4dacf3ff008a217fc10"
+  name = "github.com/lib/pq"
+  packages = [
+    ".",
+    "oid",
+  ]
+  pruneopts = ""
+  revision = "4ded0e9383f75c197b3a2aaa6d590ac52df6fd79"
+  version = "v1.0.0"
+
+[[projects]]
+  digest = "1:8bbdb2b3dce59271877770d6fe7dcbb8362438fa7d2e1e1f688e4bf2aac72706"
   name = "github.com/mattn/go-sqlite3"
   packages = ["."]
-  revision = "ed69081a91fd053f17672236b0dd52ba7485e1a3"
-  version = "v1.4.0"
+  pruneopts = ""
+  revision = "c7c4067b79cc51e6dfdcef5c702e74b1e0fa7c75"
+  version = "v1.10.0"
 
 [[projects]]
+  digest = "1:b23d258c6698b097b205b6262d8faf5a5302cc18025916cf84399b3dbf79f9ea"
+  name = "github.com/mongodb/mongo-go-driver"
+  packages = [
+    "bson",
+    "bson/bsoncodec",
+    "bson/bsonrw",
+    "bson/bsontype",
+    "bson/primitive",
+    "event",
+    "internal",
+    "mongo",
+    "mongo/options",
+    "mongo/readconcern",
+    "mongo/readpref",
+    "mongo/writeconcern",
+    "tag",
+    "version",
+    "x/bsonx",
+    "x/bsonx/bsoncore",
+    "x/mongo/driver",
+    "x/mongo/driver/auth",
+    "x/mongo/driver/auth/internal/gssapi",
+    "x/mongo/driver/session",
+    "x/mongo/driver/topology",
+    "x/mongo/driver/uuid",
+    "x/network/address",
+    "x/network/command",
+    "x/network/compressor",
+    "x/network/connection",
+    "x/network/connstring",
+    "x/network/description",
+    "x/network/result",
+    "x/network/wiremessage",
+  ]
+  pruneopts = ""
+  revision = "29905d4bda472574c8b499e2f93b5b2747d8fbd7"
+  version = "v0.1.0"
+
+[[projects]]
+  digest = "1:256484dbbcd271f9ecebc6795b2df8cad4c458dd0f5fd82a8c2fa0c29f233411"
   name = "github.com/pmezard/go-difflib"
   packages = ["difflib"]
+  pruneopts = ""
   revision = "792786c7400a136282c1664665ae0a8db921c6c2"
   version = "v1.0.0"
 
 [[projects]]
-  branch = "master"
+  digest = "1:04b59baeb0ff64fb45553afbec8b715049f3aa1dd56e31d71251dcba03feba94"
+  name = "github.com/rakanalh/scheduler"
+  packages = [
+    "storage",
+    "task",
+  ]
+  pruneopts = ""
+  revision = "3af1a0b9d9fa3c11ae657e01e2e1d22b07ea6a56"
+  version = "0.1"
+
+[[projects]]
+  digest = "1:711eebe744c0151a9d09af2315f0bb729b2ec7637ef4c410fa90a18ef74b65b6"
   name = "github.com/stretchr/objx"
   packages = ["."]
-  revision = "1a9d0bb9f541897e62256577b352fdbc1fb4fd94"
+  pruneopts = ""
+  revision = "477a77ecc69700c7cdeb1fa9e129548e1c1c393c"
+  version = "v0.1.1"
 
 [[projects]]
+  digest = "1:c587772fb8ad29ad4db67575dad25ba17a51f072ff18a22b4f0257a4d9c24f75"
   name = "github.com/stretchr/testify"
-  packages = ["assert","mock"]
-  revision = "69483b4bd14f5845b5a1e55bca19e954e827f1d0"
-  version = "v1.1.4"
+  packages = [
+    "assert",
+    "mock",
+    "require",
+  ]
+  pruneopts = ""
+  revision = "f35b8ab0b5a2cef36673838d662e249dd9c94686"
+  version = "v1.2.2"
 
 [[projects]]
   branch = "master"
-  name = "golang.org/x/net"
-  packages = ["context"]
-  revision = "9dfe39835686865bff950a07b394c12a98ddc811"
+  digest = "1:ad74f33a69bd6ab0bd7287003b7c1069b94cfb5213eb5597005fe2963d7dfca9"
+  name = "github.com/xdg/scram"
+  packages = ["."]
+  pruneopts = ""
+  revision = "7eeb5667e42c09cb51bf7b7c28aea8c56767da90"
+
+[[projects]]
+  branch = "master"
+  digest = "1:62f6eb06f6f39d18fc961956116a50a0e52b89961ed8a83bbb950b4cfb09fe67"
+  name = "github.com/xdg/stringprep"
+  packages = ["."]
+  pruneopts = ""
+  revision = "73f8eece6fdcd902c185bf651de50f3828bed5ed"
+
+[[projects]]
+  branch = "master"
+  digest = "1:887074c37fcefc2f49b5ae9c6f9f36107341aec23185613d0e9f1ee81db7f94a"
+  name = "golang.org/x/crypto"
+  packages = ["pbkdf2"]
+  pruneopts = ""
+  revision = "505ab145d0a99da450461ae2c1a9f6cd10d1f447"
+
+[[projects]]
+  branch = "master"
+  digest = "1:814474ab808c5e04b9334d046f8a0060fc1724c2c02acfd00a7cc0008d675455"
+  name = "golang.org/x/sync"
+  packages = ["semaphore"]
+  pruneopts = ""
+  revision = "37e7f081c4d4c64e13b10787722085407fe5d15f"
+
+[[projects]]
+  digest = "1:5acd3512b047305d49e8763eef7ba423901e85d5dd2fd1e71778a0ea8de10bd4"
+  name = "golang.org/x/text"
+  packages = [
+    "internal/gen",
+    "internal/triegen",
+    "internal/ucd",
+    "transform",
+    "unicode/cldr",
+    "unicode/norm",
+  ]
+  pruneopts = ""
+  revision = "f21a4dfb5e38f5895301dc265a8def02365cc3d0"
+  version = "v0.3.0"
 
 [solve-meta]
   analyzer-name = "dep"
   analyzer-version = 1
-  inputs-digest = "f6ba8bbd722c1d1ed3a92a261256f3c1ba11836e35313e2b59da977114e85a1a"
+  input-imports = [
+    "github.com/lib/pq",
+    "github.com/mattn/go-sqlite3",
+    "github.com/mongodb/mongo-go-driver/bson",
+    "github.com/mongodb/mongo-go-driver/mongo",
+    "github.com/mongodb/mongo-go-driver/x/bsonx",
+    "github.com/rakanalh/scheduler/storage",
+    "github.com/rakanalh/scheduler/task",
+    "github.com/stretchr/testify/mock",
+    "github.com/stretchr/testify/require",
+  ]
   solver-name = "gps-cdcl"
   solver-version = 1

--- a/storage/mongo.go
+++ b/storage/mongo.go
@@ -8,8 +8,8 @@ import (
 	"fmt"
 	"log"
 
-	"github.com/mongodb/mongo-go-driver/bson"
 	"github.com/mongodb/mongo-go-driver/mongo"
+	"github.com/mongodb/mongo-go-driver/x/bsonx"
 )
 
 const COLLECTION_NAME string = "task_store"
@@ -17,7 +17,7 @@ const COLLECTION_NAME string = "task_store"
 // MongoDBConfig is the config structure holding information about mongo db.
 type MongoDBConfig struct {
 	ConnectionUrl string
-	Db string
+	Db            string
 }
 
 // MongoDBStorage is the structure responsible for handling mongo storage.
@@ -41,10 +41,10 @@ func (mongodb *MongoDBStorage) Connect() error {
 	if err != nil {
 		return err
 	}
-	
+
 	mongodb.client = client
 	err = mongodb.client.Connect(context.TODO())
-	
+
 	return err
 }
 
@@ -74,7 +74,8 @@ func (mongodb MongoDBStorage) Add(task TaskAttributes) error {
 		return errors.New("could not get collection")
 	}
 
-	filter := bson.NewDocument(bson.EC.String("hash", task.Hash))
+	// filter := bson.NewDocument(bson.EC.String("hash", task.Hash))
+	filter := bsonx.Doc{{"hash", bsonx.String(task.Hash)}}
 	res, err := task_store.Count(context.Background(), filter)
 
 	if err != nil {
@@ -108,7 +109,8 @@ func (mongodb MongoDBStorage) Remove(task TaskAttributes) error {
 		return errors.New("could not get collection")
 	}
 
-	filter := bson.NewDocument(bson.EC.String("hash", task.Hash))
+	// filter := bson.NewDocument(bson.EC.String("hash", task.Hash))
+	filter := bsonx.Doc{{"hash", bsonx.String(task.Hash)}}
 
 	_, err := task_store.DeleteOne(context.Background(), filter)
 
@@ -135,7 +137,7 @@ func (mongodb MongoDBStorage) Fetch() ([]TaskAttributes, error) {
 	defer cur.Close(context.Background())
 
 	for cur.Next(context.Background()) {
-		elem := bson.NewDocument()
+		elem := bsonx.Doc{}
 		err := cur.Decode(elem)
 		if err != nil {
 			log.Println(err.Error())

--- a/storage/mongo.go
+++ b/storage/mongo.go
@@ -137,8 +137,8 @@ func (mongodb MongoDBStorage) Fetch() ([]TaskAttributes, error) {
 	defer cur.Close(context.Background())
 
 	for cur.Next(context.Background()) {
-		elem := bsonx.Doc{}
-		err := cur.Decode(elem)
+		var elem bsonx.Doc
+		err := cur.Decode(&elem)
 		if err != nil {
 			log.Println(err.Error())
 			return nil, err

--- a/storage/mongo_test.go
+++ b/storage/mongo_test.go
@@ -21,7 +21,7 @@ const pass string = ""
 
 var mongoStorage *TMongoStorage = &TMongoStorage{}
 var mongoConfig MongoDBConfig = MongoDBConfig{
-	ConnectionUrl: "mongodb://localhost/test",
+	ConnectionUrl: "mongodb://localhost:32772/test",
 	Db:            "test",
 }
 
@@ -43,8 +43,7 @@ func (s *TMongoStorage) Init(config MongoDBConfig, t *testing.T) {
 		require.NoError(t, err)
 		_, err = s.storage.client.Database(config.Db).
 			Collection(COLLECTION_NAME).
-			DeleteMany(context.Background(),
-				bsonx.Doc{})
+			DeleteMany(context.Background(), bsonx.Doc{})
 		require.NoError(t, err)
 		s.uninit = sync.Once{}
 	})
@@ -70,9 +69,10 @@ func TestFetch(t *testing.T) {
 	fetchTask := sampleTask
 	fetchTask.Hash = "C"
 
-	err := mongoStorage.storage.Add(fetchTask)
-	require.NoError(t, err)
+	errAdd := mongoStorage.storage.Add(fetchTask)
+	require.NoError(t, errAdd)
 	tasks, err := mongoStorage.storage.Fetch()
+	require.NoError(t, err)
 
 	var found bool = false
 	for _, v := range tasks {

--- a/storage/mongo_test.go
+++ b/storage/mongo_test.go
@@ -21,7 +21,7 @@ const pass string = ""
 
 var mongoStorage *TMongoStorage = &TMongoStorage{}
 var mongoConfig MongoDBConfig = MongoDBConfig{
-	ConnectionUrl: "mongodb://localhost:32772/test",
+	ConnectionUrl: "mongodb://localhost/test",
 	Db:            "test",
 }
 

--- a/storage/mongo_test.go
+++ b/storage/mongo_test.go
@@ -5,7 +5,7 @@ import (
 	"sync"
 	"testing"
 
-	"github.com/mongodb/mongo-go-driver/bson"
+	"github.com/mongodb/mongo-go-driver/x/bsonx"
 	"github.com/stretchr/testify/require"
 )
 
@@ -44,7 +44,7 @@ func (s *TMongoStorage) Init(config MongoDBConfig, t *testing.T) {
 		_, err = s.storage.client.Database(config.Db).
 			Collection(COLLECTION_NAME).
 			DeleteMany(context.Background(),
-				bson.NewDocument())
+				bsonx.Doc{})
 		require.NoError(t, err)
 		s.uninit = sync.Once{}
 	})

